### DISCUSSION TITLE
Why I can't TemplateBinding to TextDecorations?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#10696] [Q&A] [answered]
I am afraid of asking similar questions recently.

Here is my code of XAML:
![image](https://user-images.githubusercontent.com/30720036/226064303-97f71ad0-f46c-4f94-85dc-cc7ea2bce5ca.png)
PS: I am afraid Github cannot render all XAML codes correctly so I have to post it as an image.

And here is code-behind:
```
using Avalonia.Controls;
using Avalonia;
using Avalonia.Controls.Primitives;
using Avalonia.Markup.Xaml;
using Avalonia.Media;

namespace Sample.Controls
{
    public class TButton : Button
    {            
        public static readonly StyledProperty<TextDecorationCollection?> TextDecorationsProperty =
AvaloniaProperty.Register<Button, TextDecorationCollection?>(nameof(TextDecorations));
        public TextDecorationCollection? TextDecorations
        {
            get { return GetValue(TextDecorationsProperty); }
            set { SetValue(TextDecorationsProperty, value); }
        }
    }
}
```

The code of TextDecorations Dependency Property is copied from source code of TextBlock in
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/TextBlock.cs

I test my code like this:
```
<Test:TButton TextDecorations="Underline">1123123</Test:TButton>
```

However, it doesn't work any.

Soon I found a strange situation, when I replace the code 
```
<Setter Property="TextDecorations" Value="{TemplateBinding controls:TButton.TextDecorations}" />
```
to
```
<Setter Property="TextDecorations" Value="underline" />
```

Well, it works now.

It seems there is something wrong when binding in my code.

What's wrong with my code?

**Comment by amwx** [comment, marked as answer]
IIRC `TemplateBinding` doesn't work outside of `ControlTemplate`s. 

You could try: 
```xaml
<Setter Property="TextDecorations" Value="{Binding $parent[controls:TButton].TextDecorations}" />
```

**Reply by kyuranger** [reply]
> ```
> Binding $parent[controls:TButton].TextDecorations}
> ```

Well, according to your way, it works now.

Whereas, previously I bind another style like this (also outside the ControlTemplate):
`
	<Style Selector="controls|TButton:pointerover /template/ ContentPresenter#CP">		
		<Setter Property="Foreground" Value="{TemplateBinding controls:TButton.HoverForeground}" />
	</Style>
`
Just like this:
![image](https://user-images.githubusercontent.com/30720036/226108495-cc917c0a-9f6d-4bea-9489-e0a993b89da2.png)

And it works well too.

Why does the situation above works even outside ControlTemplate?

**Reply by amwx** [reply]
It could be related to the `ContentPresenter > TextBlock` part. Can you look in devtools to see what the `TemplatedParent` property is set to for that TextBlock inside the ContentPresenter. My guess is its set to `null`, as that's what I'm seeing in a normal `Button`. I'm not sure if that's a bug or not, my initial thought is that the TemplatedParent should be inherited through the entire template, but I'm not 100% sure here as that may cause issues with TemplatedControls within TemplatedControls?

**Reply by kyuranger** [reply]
> It could be related to the `ContentPresenter > TextBlock` part. Can you look in devtools to see what the `TemplatedParent` property is set to for that TextBlock inside the ContentPresenter. My guess is its set to `null`, as that's what I'm seeing in a normal `Button`. I'm not sure if that's a bug or not, my initial thought is that the TemplatedParent should be inherited through the entire template, but I'm not 100% sure here as that may cause issues with TemplatedControls within TemplatedControls?

I just tested it. TemplatedParent of the TextBlock is null indeed. Thank you for telling me for a way to get the TemplatedParent in devtools.
